File: src/main/cs/RLBotDotNet/RLBotDotNet/GameState/DesiredVector3.cs
using rlbot.flat;
using FlatBuffers;

namespace RLBotDotNet.GameState
{
    public class DesiredVector3
    {
        public float? X;
        public float? Y;
        public float? Z;

        public static DesiredVector3 Zero => new DesiredVector3(0, 0, 0);

        public DesiredVector3(float? x = null, float? y = null, float? z = null)
        {
            X = x;
            Y = y;
            Z = z;
        }

        public DesiredVector3(Vector3 vector)
        {
            X = vector.X;
            Y = vector.Y;
            Z = vector.Z;
        }

        public DesiredVector3(System.Numerics.Vector3 vector)
        {
            X = vector.X;
            Y = vector.Y;
            Z = vector.Z;
        }

        public Offset<Vector3Partial> ToFlatBuffer(FlatBufferBuilder builder)
        {
            Vector3Partial.StartVector3Partial(builder);

            if (X.HasValue)
                Vector3Partial.AddX(builder, Float.CreateFloat(builder, X.Value));

            if (Y.HasValue)
                Vector3Partial.AddY(builder, Float.CreateFloat(builder, Y.Value));

            if (Z.HasValue)
                Vector3Partial.AddZ(builder, Float.CreateFloat(builder, Z.Value));

            return Vector3Partial.EndVector3Partial(builder);
        }
    }
}

File: src/main/java/rlbot/gamestate/DesiredVector3.java
package rlbot.gamestate;

import com.google.flatbuffers.FlatBufferBuilder;
import rlbot.flat.Vector3Partial;

/**
 * See https://github.com/RLBot/RLBotJavaExample/wiki/Manipulating-Game-State
 */
public class DesiredVector3 {
    private Float x;
    private Float y;
    private Float z;

    public DesiredVector3() {
    }

    public DesiredVector3(Float x, Float y, Float z) {
        this.x = x;
        this.y = y;
        this.z = z;
    }

    public Float getX() {
        return x;
    }
    
    /**
     * @returns {@code true} If x is not null.
     */
    public boolean hasX() {
        return x != null
    }

    public DesiredVector3 withX(Float x) {
        this.x = x;
        return this;
    }

    public Float getY() {
        return y;
    }
    
    /**
     * @returns {@code true} If y is not null.
     */
    public boolean hasY() {
        return y != null
    }

    public DesiredVector3 withY(Float y) {
        this.y = y;
        return this;
    }

    public Float getZ() {
        return z;
    }
    
    /**
     * @returns {@code true} If z is not null.
     */
    public boolean hasZ() {
        return z != null
    }

    public DesiredVector3 withZ(Float z) {
        this.z = z;
        return this;
    }

    /**
     * You may wish to override this in order to invert the x value, which would give you a normal
     * right-handed coordinate system.
     */
    public int toFlatbuffer(FlatBufferBuilder builder) {
        Vector3Partial.startVector3Partial(builder);
        if (hasX()) {
            Vector3Partial.addX(builder, rlbot.flat.Float.createFloat(builder, getX()));
        }
        if (hasY()) {
            Vector3Partial.addY(builder, rlbot.flat.Float.createFloat(builder, getY()));
        }
        if (hasZ()) {
            Vector3Partial.addZ(builder, rlbot.flat.Float.createFloat(builder, getZ()));
        }
        return Vector3Partial.endVector3Partial(builder);
    }
}
